File: app/services/task_suggestion_service.py
"""ML-подсказки полей задачи: labels, skills, difficulty и deadline по title/description."""

from datetime import UTC, datetime, timedelta

from app.integrations.openrouter_client import SuggestionCandidate, TaskSuggestionSource
from app.repositories.skill_repo import SkillRepository
from app.repositories.task_repo import TaskRepository
from app.schemas.experience import TaskSkillRead
from app.schemas.task_suggestion import (
    TaskAnalyzeRequest,
    TaskFieldSuggestion,
)
from app.services.exceptions import ServiceUnavailableError

_DEFAULT_DIFFICULTY = 3
_SKILL_CANDIDATE_PAGE_SIZE = 1000


def _clamp(value: int, low: int, high: int) -> int:
    return max(low, min(high, value))


class TaskSuggestionService:
    """Запрашивает подсказки у ML-провайдера и валидирует их против данных БД.

    Ответу модели не доверяем: несуществующие id отбрасываются,
    difficulty и exp_reward зажимаются в допустимые диапазоны.
    """

    def __init__(
        self,
        task_repo: TaskRepository,
        skill_repo: SkillRepository,
        source: TaskSuggestionSource | None,
    ) -> None:
        self.task_repo = task_repo
        self.skill_repo = skill_repo
        self.source = source

    async def analyze(self, data: TaskAnalyzeRequest) -> TaskFieldSuggestion:
        if self.source is None:
            raise ServiceUnavailableError("ML suggestions are unavailable: set OPENROUTER_API_KEY")

        labels = await self.task_repo.list_labels()
        skills = []
        offset = 0
        while True:
            page = await self.skill_repo.list(limit=_SKILL_CANDIDATE_PAGE_SIZE, offset=offset)
            skills.extend(page)
            if len(page) < _SKILL_CANDIDATE_PAGE_SIZE:
                break
            offset += len(page)

        raw = await self.source.suggest_fields(
            title=data.title,
            description=data.description,
            labels=[SuggestionCandidate(id=label.id, name=label.name) for label in labels],
            skills=[SuggestionCandidate(id=skill.id, name=skill.name) for skill in skills],
        )

        labels_by_id = {label.id: label for label in labels}
        skills_by_id = {skill.id: skill for skill in skills}

        difficulty = _DEFAULT_DIFFICULTY
        if raw.difficulty is not None:
            difficulty = _clamp(raw.difficulty, 1, 5)

        # Модель оценивает срок в днях; в конкретную дату переводим здесь,
        # потому что LLM не знает текущей даты.
        deadline = None
        if raw.estimated_days is not None:
            days = _clamp(raw.estimated_days, 1, 365)
            deadline = datetime.now(UTC).replace(microsecond=0) + timedelta(days=days)

        suggested_labels = [
            labels_by_id[label_id] for label_id in raw.label_ids if label_id in labels_by_id
        ]

        seen_skill_ids: set[int] = set()
        suggested_skills: list[TaskSkillRead] = []
        for skill_id, exp_reward in raw.skills:
            skill = skills_by_id.get(skill_id)
            if skill is None or skill_id in seen_skill_ids:
                continue
            seen_skill_ids.add(skill_id)
            suggested_skills.append(
                TaskSkillRead(
                    skill=skill,
                    exp_reward=_clamp(exp_reward, 1, 1000),
                )
            )

        return TaskFieldSuggestion(
            difficulty=difficulty,
            deadline=deadline,
            labels=suggested_labels,
            skills=suggested_skills,
        )

File: backend/src/main/java/com/skilltracker/service/TaskSuggestionService.java
package com.skilltracker.service;

import com.skilltracker.domain.Label;
import com.skilltracker.domain.Skill;
import com.skilltracker.dto.LabelResponse;
import com.skilltracker.dto.SkillResponse;
import com.skilltracker.dto.TaskAnalyzeRequest;
import com.skilltracker.dto.TaskFieldSuggestionResponse;
import com.skilltracker.dto.TaskSkillResponse;
import com.skilltracker.exception.ServiceUnavailableException;
import com.skilltracker.integration.RawTaskSuggestion;
import com.skilltracker.integration.SuggestionCandidate;
import com.skilltracker.integration.TaskSuggestionSource;
import com.skilltracker.repository.LabelRepository;
import com.skilltracker.repository.OffsetLimit;
import com.skilltracker.repository.SkillRepository;
import java.time.Clock;
import java.time.OffsetDateTime;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Turns a model suggestion into a response that only ever references rows that exist.
 *
 * <p>The model output is untrusted: unknown ids are dropped and every numeric field is clamped to
 * the range the API accepts.
 */
@Service
public class TaskSuggestionService {

    private static final int DEFAULT_DIFFICULTY = 3;
    private static final int SKILL_CANDIDATE_PAGE_SIZE = 1000;

    private final LabelRepository labelRepository;
    private final SkillRepository skillRepository;
    private final Optional<TaskSuggestionSource> source;
    private final Clock clock;

    public TaskSuggestionService(
            LabelRepository labelRepository,
            SkillRepository skillRepository,
            Optional<TaskSuggestionSource> source,
            Clock clock) {
        this.labelRepository = labelRepository;
        this.skillRepository = skillRepository;
        this.source = source;
        this.clock = clock;
    }

    @Transactional(readOnly = true)
    public TaskFieldSuggestionResponse analyze(TaskAnalyzeRequest request) {
        TaskSuggestionSource suggestionSource = source.orElseThrow(
                () -> new ServiceUnavailableException("ML suggestions are unavailable: set OPENROUTER_API_KEY"));

        List<Label> labels = labelRepository.findAllByOrderByNameAsc();
        List<Skill> skills = loadAllSkills();

        RawTaskSuggestion raw = suggestionSource.suggestFields(
                request.title(),
                request.description(),
                labels.stream()
                        .map(label -> new SuggestionCandidate(label.getId(), label.getName()))
                        .toList(),
                skills.stream()
                        .map(skill -> new SuggestionCandidate(skill.getId(), skill.getName()))
                        .toList());

        Map<Integer, Label> labelsById = new LinkedHashMap<>();
        labels.forEach(label -> labelsById.put(label.getId(), label));
        Map<Integer, Skill> skillsById = new LinkedHashMap<>();
        skills.forEach(skill -> skillsById.put(skill.getId(), skill));

        return new TaskFieldSuggestionResponse(
                raw.difficulty() == null ? DEFAULT_DIFFICULTY : clamp(raw.difficulty(), 1, 5),
                deadlineFor(raw.estimatedDays()),
                suggestedLabels(raw, labelsById),
                suggestedSkills(raw, skillsById));
    }

    /**
     * The model has no idea what today is, so it estimates an effort in days and the deadline is
     * derived here.
     */
    private OffsetDateTime deadlineFor(Integer estimatedDays) {
        if (estimatedDays == null) {
            return null;
        }
        return OffsetDateTime.now(clock).truncatedTo(ChronoUnit.SECONDS).plusDays(clamp(estimatedDays, 1, 365));
    }

    private List<LabelResponse> suggestedLabels(RawTaskSuggestion raw, Map<Integer, Label> labelsById) {
        return raw.labelIds().stream()
                .filter(labelsById::containsKey)
                .map(labelId -> LabelResponse.from(labelsById.get(labelId)))
                .toList();
    }

    private List<TaskSkillResponse> suggestedSkills(RawTaskSuggestion raw, Map<Integer, Skill> skillsById) {
        Set<Integer> seen = new HashSet<>();
        List<TaskSkillResponse> suggested = new ArrayList<>();
        for (RawTaskSuggestion.SkillReward reward : raw.skills()) {
            Skill skill = skillsById.get(reward.skillId());
            if (skill == null || !seen.add(reward.skillId())) {
                continue;
            }
            suggested.add(new TaskSkillResponse(SkillResponse.from(skill), clamp(reward.expReward(), 1, 1000)));
        }
        return suggested;
    }

    private List<Skill> loadAllSkills() {
        List<Skill> skills = new ArrayList<>();
        int offset = 0;
        while (true) {
            List<Skill> page = skillRepository.findAllOrderedById(
                    OffsetLimit.of(SKILL_CANDIDATE_PAGE_SIZE, offset, Sort.unsorted()));
            skills.addAll(page);
            if (page.size() < SKILL_CANDIDATE_PAGE_SIZE) {
                return skills;
            }
            offset += page.size();
        }
    }

    private int clamp(int value, int low, int high) {
        return Math.max(low, Math.min(high, value));
    }
}
